pytest-embedded/pytest_embedded/dut_factory.py

```diff
@@ -0,0 +1,655 @@
+    qemu: t.Optional['Qemu'],
+    wokwi: t.Optional['WokwiCLI'],
+) -> t.Union[Dut, t.List[Dut]]:
+    global dut_index_global
+    dut_index_global += 1
+
+    if hasattr(app, 'target') and app.target == 'linux':
+        from pytest_embedded_idf import LinuxDut
+
+        cls = LinuxDut
+        kwargs = _fixture_classes_and_options.kwargs['dut']
+        kwargs['serial'] = serial
+
+        return cls(**kwargs)
+
+    cls = _fixture_classes_and_options.classes['dut']
+    mixins = _fixture_classes_and_options.mixins['dut']
+    kwargs = _fixture_classes_and_options.kwargs['dut']
+
+    for k, v in kwargs.items():
+        if v is None:
+            if k == 'app':
+                kwargs[k] = app
+            elif k == 'serial':
+                kwargs[k] = serial
+            elif k == 'openocd':
+                kwargs[k] = openocd
+            elif k == 'gdb':
+                kwargs[k] = gdb
+            elif k == 'qemu':
+                kwargs[k] = qemu
+            elif k == 'wokwi':
+                kwargs[k] = wokwi
+    return cls(**_drop_none_kwargs(kwargs), mixins=mixins)
+
+
+buff_parametrized_fixtures = {}
+
+
+def set_buff_parametrized_fixtures(values: t.Dict):
+    global buff_parametrized_fixtures
+    buff_parametrized_fixtures = values.copy()
+
+
+def _close_or_terminate(obj):
+    if obj is None:
+        del obj
+        return
+
+    try:
+        if isinstance(obj, (subprocess.Popen, multiprocessing.process.BaseProcess)):
+            obj.terminate()
+            obj.kill()
+        elif isinstance(obj, io.IOBase):
+            try:
+                obj.close()
+            except Exception as e:
+                logging.debug('file %s closed failed with error: %s', obj, str(e))
+        else:
+            try:
+                obj.close()
+            except AttributeError:
+                try:
+                    obj.terminate()
+                except AttributeError:
+                    pass
+            except Exception as e:
+                logging.debug('Not properly caught object %s: %s', obj, str(e))
+    except Exception as e:
+        logging.debug('%s: %s', obj, str(e))
+        return  # swallow up all error
+    finally:
+        referrers = gc.get_referrers(obj)
+        for _referrer in referrers:
+            if isinstance(_referrer, list):
+                for _i, val in enumerate(_referrer):
+                    if val is obj:
+                        _referrer[_i] = None
+            elif isinstance(_referrer, dict):
+                for key, value in _referrer.items():
+                    if value is obj:
+                        _referrer[key] = None
+        del obj
+
+
+class DutFactory:
+    def __init__(self):
+        self.obj_stack = []
+
+    # The object has to be closed;
+    # otherwise, the process will still be here,
+    # but there are some problems with deletion, most likely because of the fork process.
+    # Call close instead delete.
+    def close(self):
```

Comment:
At first, I tried using __del__, but it seems there are some issues with the process. So changed to close